modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/metadata/ColocationGroup.java

```diff
@@ -230,29 +300,30 @@ public ColocationGroup mapToNodes(List<String> nodeNames) {
 
     @NotNull
     private ColocationGroup forNodes0(List<String> nodeNames) {
-        List<List<String>> assignments = new ArrayList<>(SYNTHETIC_PARTITIONS_COUNT);
-        for (int i = 0; i < SYNTHETIC_PARTITIONS_COUNT; i++) {
-            assignments.add(asList(nodeNames.get(i % nodeNames.size())));
-        }
         return new ColocationGroup(sourceIds, nodeNames, assignments);
     }
 
     /**
-     * Returns List of partitions to scan on the given node.
+     * Returns list of pairs containing the partition number to scan on the given node with the corresponding primary replica term.
      *
-     * @param nodeNames Cluster node consistent ID.
-     * @return List of partitions to scan on the given node.
+     * @param nodeName Cluster node consistent ID.
+     * @return List of pairs containing the partition number to scan on the given node with the corresponding primary replica term.
      */
-    public int[] partitions(String nodeNames) {
-        IgniteIntList parts = new IgniteIntList(assignments.size());
+    public List<PartitionWithTerm> partitionsWithTerms(String nodeName) {
+        List<PartitionWithTerm> partsWithTerms = new ArrayList<>();
```

Comment:
```suggestion
        List<PartitionWithTerm> partsWithTerms = new ArrayList<>(assignments.size());
```

Reply:
I'm not sure if this is a good improvement, because it will most likely result in a constant small memory leak. In most cases, only a some part of the partitions belongs to a specified node.